slapo/schedule.py

```diff
@@ -276,109 +249,165 @@ def set_output_type(output_type, gather_axis=None):
                         f"{gather_axis} is requested"
                     ) from None
 
-        # Update attributes. FIXME: Generalize to other ops.
-        if isinstance(self.mod, nn.Linear):
-            if axis == 0:
-                self.mod.out_features = sharded_size
-                # Note that the axis is the axis of the output
-                set_output_type("partition", gather_axis=1)
-            else:  # axis == 1
-                self.mod.in_features = sharded_size
-                set_output_type("partial")
-        elif isinstance(self.mod, nn.Conv2d):
-            axes = [1, 0] if self.mod.transposed else [0, 1]
-            if axis == axes[0]:
-                self.mod.out_channels = sharded_size
-                set_output_type("partition", gather_axis=1)
-            elif axis == axes[1]:
-                self.mod.in_channels = sharded_size
-                set_output_type("partial")
-            else:
-                raise NotImplementedError
-        elif isinstance(self.mod, nn.BatchNorm2d):
-            self.mod.num_features = sharded_size
-            set_output_type("partition", gather_axis=1)
+        out_type, out_part_axis = get_output_type_after_sharding(
+            self.mod, sharded_size, axis
+        )
+        if out_type is not None:
+            set_output_type(out_type, gather_axis=out_part_axis)
 
     @register_primitive(need_dist=True)
-    def sync(self, mode="backward"):
-        """There are several cases for sync based on two factors:
-        1) The original forward output is partitioned or partial sum.
-        2) The next module wants to take full or partitioned input.
-        Note that we ignore the case that the next module wants to take partial sum
-        input, because it is not benefitical to the performance.
+    def sync(self, mode, sync_op_or_fn, **kwargs):
+        """Synchronize the tensor across multiple devices.
+        Since the underlying implementation is registering a PyTorch hook
+        to the target module, the mode could be "fwd_pre", "fwd_post", "bwd_post".
+        The following are some example use cases:
 
         Case 1: (replica x, shard_out w) -> partition output -> allgather
                 -> full output -> (replica x, shard_out w).
             In this case, since forward uses all-gather to get a full output,
             backward must have a split to match the shape, and
-            allreduce is also required for x.grad, so mode should be 'both'.
+            allreduce is also required for x.grad, so we use:
+            ```python
+            sch["out_prj"].shard("weight", axis=0)
+            sch["out_prj"].sync(mode="fwd_post", sync_op_or_fn="all_gather", axis=1)
```

Comment:
i found it might a bit confusing since shard(axis=0) is for weight output dimension but sync(axis=1) for activation output dimension. maybe add a bit more text to explain?